File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/api/model/activity/Polar247HrSamplesData.kt
package com.polar.sdk.api.model.activity

import com.polar.androidcommunications.api.ble.BleLogger
import com.polar.sdk.impl.utils.PolarTimeUtils
import fi.polar.remote.representation.protobuf.AutomaticSamples
import fi.polar.remote.representation.protobuf.AutomaticSamples.PbAutomaticSampleSessions
import java.time.LocalDate
import java.time.LocalTime

private const val TAG = "Polar247HrSamplesData"

data class Polar247HrSamplesData(
    val date: LocalDate,
    val samples: List<Polar247HrSamples>
) {
    companion object {
        fun fromProto(proto: PbAutomaticSampleSessions): Polar247HrSamplesData {
            return Polar247HrSamplesData(
                PolarTimeUtils.pbDateToLocalDate(proto.day),
                Polar247HrSamples.fromProto(proto.samplesList)
            )
        }
    }
}

data class Polar247HrSamples(
    val startTime: LocalTime,
    val hrSamples: List<Int>,
    val triggerType: AutomaticSampleTriggerType
) {
    companion object {
        fun fromProto(heartRateSampleProtos: List<AutomaticSamples.PbAutomaticHeartRateSamples>): List<Polar247HrSamples> {
            var heartRateSampleList = mutableListOf<Polar247HrSamples>()
            for (heartRateSampleProto in heartRateSampleProtos ) {
                val triggerTypeValue = heartRateSampleProto.triggerType.number
                val triggerType = AutomaticSampleTriggerType.fromProto(triggerTypeValue)
                if (triggerType == null) {
                    BleLogger.w(TAG, "Skipping HR sample with unknown trigger type: $triggerTypeValue")
                    continue
                }
                val sample = Polar247HrSamples(
                    PolarTimeUtils.pbTimeToLocalTime(heartRateSampleProto.time),
                    heartRateSampleProto.heartRateList,
                    triggerType
                )
                heartRateSampleList.add(sample)
            }
            return heartRateSampleList
        }
    }
}

enum class AutomaticSampleTriggerType(val value: Int) {
    /**
     * Automatic measurement triggered by user high activity. Contains 1-n samples
     */
    TRIGGER_TYPE_HIGH_ACTIVITY(1),

    /**
     * Automatic measurement triggered by user low activity. Contains 1-n samples
     */
    TRIGGER_TYPE_LOW_ACTIVITY(2),

    /**
     * Automatic measurement triggered by timer. Contains 1 sample
     */
    TRIGGER_TYPE_TIMED(3),

    /**
     * Manual measurement (other than exercise) triggered by user. Contains 1-n samples
     */
    TRIGGER_TYPE_MANUAL(4);

    companion object {
        /**
         * Returns the matching [AutomaticSampleTriggerType] for the given protobuf number,
         * or **null** if the device reports an unknown trigger type (e.g. from newer firmware).
         * Callers should skip samples with a null trigger type and log the unknown value.
         */
        infix fun fromProto(value: Int): AutomaticSampleTriggerType? =
            AutomaticSampleTriggerType.values().firstOrNull { it.value == value }
    }
}

File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/api/model/activity/Polar247HrSamplesData.swift
///  Copyright © 2024 Polar. All rights reserved.

import Foundation

public struct Polar247HrSamplesData: Codable {
    
    public let date: DateComponents
    public let samples: [Polar247HrSample]
    
    public struct Polar247HrSample: Codable {

        public var time: DateComponents
        public var hrSamples: [UInt32]
        public var triggerType: AutomaticSampleTriggerType?
    }

    public enum AutomaticSampleTriggerType: String, Codable {
        case highActivity = "highActivity"
        case lowActivity = "lowActivity"
        case timed = "timed"
        case manual = "manual"
        
        init?(stringValue: String) {
            switch stringValue {
            case "highActivity":
                self = .highActivity
            case "lowActivity":
                self = .lowActivity
            case "timed":
                self = .timed
            case "manual":
                self = .manual
            default:
                return nil
            }
        }

        var stringValue: String {
            return self.rawValue
        }

        static func getByValue(value: Data_PbMeasTriggerType) -> AutomaticSampleTriggerType {

            switch value {
                case .triggerTypeHighActivity: return .highActivity
                case .triggerTypeLowActivity: return .lowActivity
                case .triggerTypeTimed: return .timed
                case .triggerTypeManual: return .manual
            }
        }
    }
    
    static func fromPbHrDataSamples(samples: [Data_PbAutomaticHeartRateSamples]) throws -> [Polar247HrSample] {

        var hrSamples = [Polar247HrSample]()
        var time = DateComponents()
        var triggerType = AutomaticSampleTriggerType.manual
        for sample in samples {
            do {
                hrSamples.append(try .init(time: PolarTimeUtils.pbTimeToDateComponents(pbTime: sample.time), hrSamples: sample.heartRate, triggerType: AutomaticSampleTriggerType.getByValue(value: sample.triggerType)))
            } catch let error {
                BleLogger.error("Polar247HrSamplesData.fromPbHrDataSamples failed to parse time from \(sample.time) with error: \(error)")
                throw error
            }
        }
        return hrSamples
    }
}
